fideslib/core/config.py

```diff
@@ -0,0 +1,272 @@
+import hashlib
+import logging
+import os
+from pathlib import Path
+from typing import Any, Dict, List, MutableMapping, Optional, Tuple, Type, Union
+
+import bcrypt
+import tomli
+from pydantic import AnyHttpUrl, BaseSettings, PostgresDsn, ValidationError, validator
+from pydantic.env_settings import SettingsSourceCallable
+
+from fideslib.exceptions import MissingConfig
+from fideslib.logging.logger import NotPii
+
+logger = logging.getLogger(__name__)
+
+
+class FidesSettings(BaseSettings):
+    """Class used as a base model for configuration subsections."""
+
+    class Config:  # pylint: disable=C0115
+        # Need to allow extras because the inheriting class will have more info
+        extras = "allow"
+
+        @classmethod
+        def customise_sources(
+            cls,
+            init_settings: SettingsSourceCallable,
+            env_settings: SettingsSourceCallable,
+            file_secret_settings: SettingsSourceCallable,  # pylint: disable=W0613
+        ) -> Tuple[SettingsSourceCallable, ...]:
+            """Set environment variables to take precedence over init values."""
+            return env_settings, init_settings
+
+
+class DatabaseSettings(FidesSettings):
+    """Configuration settings for Postgres."""
+
+    SERVER: str
+    USER: str
+    PASSWORD: str
+    DB: str = "test"
+    PORT: str = "5432"
+    TEST_DB: str = "test"
+
+    SQLALCHEMY_DATABASE_URI: Optional[PostgresDsn] = None
+    SQLALCHEMY_TEST_DATABASE_URI: Optional[PostgresDsn] = None
+
+    @validator("SQLALCHEMY_DATABASE_URI", pre=True)
+    def assemble_db_connection(  # pylint: disable=E0213, R0201
+        cls, v: Optional[str], values: Dict[str, str]
+    ) -> str:
+        """Join DB connection credentials into a connection string"""
+        if isinstance(v, str):
+            return v
+        return PostgresDsn.build(
+            scheme="postgresql",
+            user=values["USER"],
+            password=values["PASSWORD"],
+            host=values["SERVER"],
+            port=values.get("PORT"),
+            path=f"/{values.get('DB') or ''}",
+        )
+
+    @validator("SQLALCHEMY_TEST_DATABASE_URI", pre=True)
+    def assemble_test_db_connection(  # pylint: disable=E0213, R0201
+        cls, v: Optional[str], values: Dict[str, str]
+    ) -> str:
+        """Join DB connection credentials into a connection string"""
+        if isinstance(v, str):
+            return v
+        return PostgresDsn.build(
+            scheme="postgresql",
+            user=values["USER"],
+            password=values["PASSWORD"],
+            host=values["SERVER"],
+            port=values["PORT"],
+            path=f"/{values.get('TEST_DB') or ''}",
+        )
+
+    class Config:  # pylint: disable=C0115
+        env_prefix = "FIDES__DATABASE__"
```

Comment:
I believe it is because the of the override. I can see the potential where there could be an issue on the initial load before the config knows `FIDES__DATABASE__` has changed to something like `FIDESOPS__DATABASE__`. I will test this scenario out to see what happens.